File: diagnosis/evidence_graph/stage_window.py
"""Build process-local semantic windows from RuntimeEvent evidence."""

from __future__ import annotations

from dataclasses import dataclass
from itertools import groupby
from typing import Iterable

from diagnosis.schema import NormalizedEvent


@dataclass(frozen=True)
class StageWindow:
    window_id: str
    trace_id: str
    sequence_id: int
    stage: str
    source_node: str
    pid: int
    tids: tuple[int, ...]
    host_id: str
    clock_id: str
    start_ns: int
    end_ns: int
    start_event_id: str
    end_event_id: str

    def contains(self, timestamp_ns: int) -> bool:
        return self.start_ns <= timestamp_ns <= self.end_ns


def build_stage_windows(events: Iterable[NormalizedEvent]) -> list[StageWindow]:
    runtime_events = [event for event in events if event.source == "runtime_event"]
    for event in runtime_events:
        if not event.trace_id or not event.stage or event.pid <= 0 or event.tid <= 0:
            raise ValueError(f"incomplete RuntimeEvent identity: {event.event_id}")

    key = lambda event: (
        event.trace_id,
        event.sequence_id,
        event.pid,
        event.host_id,
        event.clock_id,
    )
    windows: list[StageWindow] = []
    for _, grouped in groupby(sorted(runtime_events, key=lambda item: (*key(item), item.timestamp_ns)), key=key):
        process_events = list(grouped)
        for index, start_event in enumerate(process_events):
            next_event = process_events[index + 1] if index + 1 < len(process_events) else None
            duration_ns = start_event.attributes.get("duration_ns", 0)
            if isinstance(duration_ns, bool) or not isinstance(duration_ns, int) or duration_ns < 0:
                raise ValueError(f"invalid duration_ns: {start_event.event_id}")
            end_ns = (
                next_event.timestamp_ns
                if next_event is not None
                else start_event.timestamp_ns + duration_ns
            )
            if end_ns < start_event.timestamp_ns:
                raise ValueError(f"non-monotonic stage events: {start_event.trace_id}")
            windows.append(
                StageWindow(
                    window_id=f"stage-window:{start_event.event_id}",
                    trace_id=start_event.trace_id,
                    sequence_id=start_event.sequence_id,
                    stage=start_event.stage,
                    source_node=str(start_event.attributes.get("source_node", "")),
                    pid=start_event.pid,
                    tids=(start_event.tid,),
                    host_id=start_event.host_id,
                    clock_id=start_event.clock_id,
                    start_ns=start_event.timestamp_ns,
                    end_ns=end_ns,
                    start_event_id=start_event.event_id,
                    end_event_id=(
                        next_event.event_id if next_event is not None else start_event.event_id
                    ),
                )
            )
    return sorted(windows, key=lambda item: (item.start_ns, item.window_id))

File: cpp_core/diagnosis/src/stage_window.cpp
#include "robotraceopt/diagnosis/stage_window.hpp"

#include <algorithm>
#include <map>
#include <limits>
#include <set>
#include <stdexcept>
#include <tuple>
#include <utility>

namespace robotraceopt::diagnosis {
namespace {

using GroupKey =
    std::tuple<std::string, std::int64_t, std::int64_t, std::string, std::string>;

std::string scalar_to_string(const AttributeValue& value) {
    if (const auto* text = std::get_if<std::string>(&value)) {
        return *text;
    }
    if (const auto* integer = std::get_if<std::int64_t>(&value)) {
        return std::to_string(*integer);
    }
    if (const auto* real = std::get_if<double>(&value)) {
        return std::to_string(*real);
    }
    if (const auto* flag = std::get_if<bool>(&value)) {
        return *flag ? "True" : "False";
    }
    return "";
}

std::int64_t duration_for(const NormalizedEvent& event) {
    const auto found = event.attributes.find("duration_ns");
    if (found == event.attributes.end()) {
        return 0;
    }
    const auto* duration = std::get_if<std::int64_t>(&found->second);
    if (duration == nullptr || *duration < 0) {
        throw std::invalid_argument("invalid duration_ns: " + event.event_id);
    }
    return *duration;
}

}  // namespace

std::vector<StageWindow> build_stage_windows(const std::vector<NormalizedEvent>& events) {
    std::map<GroupKey, std::vector<const NormalizedEvent*>> groups;
    std::set<std::string> event_ids;

    for (const auto& event : events) {
        if (event.source != "runtime_event") {
            continue;
        }
        if (event.event_id.empty()) {
            throw std::invalid_argument("RuntimeEvent event_id must not be empty");
        }
        if (!event_ids.emplace(event.event_id).second) {
            throw std::invalid_argument("duplicate RuntimeEvent event_id: " + event.event_id);
        }
        if (event.trace_id.empty() || event.stage.empty() || event.pid <= 0 || event.tid <= 0) {
            throw std::invalid_argument("incomplete RuntimeEvent identity: " + event.event_id);
        }
        groups[{event.trace_id, event.sequence_id, event.pid, event.host_id, event.clock_id}]
            .push_back(&event);
    }

    std::vector<StageWindow> windows;
    for (auto& [key, grouped] : groups) {
        (void)key;
        std::stable_sort(grouped.begin(), grouped.end(), [](const auto* left, const auto* right) {
            return left->timestamp_ns < right->timestamp_ns;
        });
        for (std::size_t index = 0; index < grouped.size(); ++index) {
            const auto& start = *grouped[index];
            const auto* next = index + 1 < grouped.size() ? grouped[index + 1] : nullptr;
            const auto duration = duration_for(start);
            std::int64_t end_ns = next == nullptr ? start.timestamp_ns : next->timestamp_ns;
            if (next == nullptr) {
                if (duration > 0 &&
                    start.timestamp_ns > std::numeric_limits<std::int64_t>::max() - duration) {
                    throw std::invalid_argument("duration_ns overflows timestamp: " + start.event_id);
                }
                end_ns += duration;
            }
            if (end_ns < start.timestamp_ns) {
                throw std::invalid_argument("non-monotonic stage events: " + start.trace_id);
            }
            std::string source_node;
            if (const auto found = start.attributes.find("source_node");
                found != start.attributes.end()) {
                source_node = scalar_to_string(found->second);
            }
            windows.push_back(StageWindow{
                "stage-window:" + start.event_id,
                start.trace_id,
                start.sequence_id,
                start.stage,
                std::move(source_node),
                start.pid,
                {start.tid},
                start.host_id,
                start.clock_id,
                start.timestamp_ns,
                end_ns,
                start.event_id,
                next == nullptr ? start.event_id : next->event_id,
            });
        }
    }
    std::sort(windows.begin(), windows.end(), [](const auto& left, const auto& right) {
        return std::tie(left.start_ns, left.window_id) < std::tie(right.start_ns, right.window_id);
    });
    return windows;
}

}  // namespace robotraceopt::diagnosis
